### PR TITLE
[devtools-blog] How we built the Chrome DevTools Issues tab - DevTools engineering blog

### PR DESCRIPTION
**Pull request by chybie**
**Target Live Date:** 2020-08-29

- [x] This has been reviewed and approved by (@chybie)
- [x] I have run `npm test` locally and all tests pass.
- [x] I have added the appropriate `type-something` label.
- [x] I've staged the site and manually verified that my content displays correctly.

**CC:** @petele, @mathiasbynens, @sigurdschneider, @jschfflr, @samdutton

### CODE REVIEW

File: src/content/en/tools/chrome-devtools/blog/index.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+book_path: /web/tools/chrome-devtools/_book.yaml
+description: Chrome DevTools Engineering Blog
+
+{# wf_blink_components: Platform>DevTools #}
+{# wf_updated_on: 2020-08-24 #}
+{# wf_published_on: 2020-07-10 #}
+
+{% include "web/tools/chrome-devtools/_shared/styles.html" %}
+
+# Engineering Blog {: .page-title }
```
Comment (mathiasbynens) [inline src/content/en/tools/chrome-devtools/blog/index.md:11]:
```suggestion
# Engineering blog {: .page-title }
```

File: src/content/en/updates/2020/08/chrome-devtools-issues-tab.md [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+The console was full of warnings and error messages, that contained rather technical explanations and sometimes links to [chromestatus.com](https://chromestatus.com). All messages looked roughly equally important, making it hard to figure out **which to address first**. More importantly, the text was not linking to additional information inside DevTools, making it difficult to understand **what happened**. And finally, the messages often left it entirely to the web developer to figure out **how to fix the problem** or even learn about the technical context.
+
+If you also use the console for messages from your own application, you’ll sometimes have a hard time finding them between all the messages from the browser.
+
+As well as humans, it's also difficult for automated processes to interact with console messages. For example, developers might use Chrome Headless and Puppeteer in a Continuous Integration/Continuous Deployment scenario. Because console messages are just strings, developers need to write regular expressions or some other parser to extract actionable information. 
+
+
+## The solution: Structured and actionable issue reporting
```
Comment (mathiasbynens) [inline src/content/en/updates/2020/08/chrome-devtools-issues-tab.md:30]:
```suggestion
## The solution: structured and actionable issue reporting
```

File: src/content/en/updates/2020/08/chrome-devtools-issues-tab.md [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+To find a better solution to the problems we discovered, we first started thinking about the requirements and collected them in a [Design Doc](https://docs.google.com/document/d/1F6R5Bpb3qHNzGPNBSXwEJ_eP8L-anIj0WinxOIyAh54/edit#heading=h.ronwk0bge64o).
```
Comment (mathiasbynens) [inline src/content/en/updates/2020/08/chrome-devtools-issues-tab.md:32]:
```suggestion
To find a better solution to the problems we discovered, we first started thinking about the requirements and collected them in a [Design Doc](https://docs.google.com/document/d/1F6R5Bpb3qHNzGPNBSXwEJ_eP8L-anIj0WinxOIyAh54/edit).
```

File: src/content/en/updates/2020/08/chrome-devtools-issues-tab.md [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+![Prototypes](/web/updates/images/2020/08/chrome-devtools-issues-tab/prototypes.png)
+
+After a lot of discussion around the best solution, we started implementing the design and reiterating decisions to gradually arrive at what the Issues tablooks like today.
+
+Another very important factor was the **discoverability** of the Issues tab. In the past, many great Devtools features were not discoverable without the developer knowing what specifically to look for. For the Issues tab, we decided to highlight issues in multiple different areas to increase the likelihood of developers opening it. The most prominent highlight is probably in the Console panel. But we also added an icon to the warnings and errors counter in the top right of the DevTools window. Finally, the Issues tab not only links to other DevTools panels, but resources that are related to an issue also link back to the Issues tab.
+
+![Related issues](/web/updates/images/2020/08/chrome-devtools-issues-tab/related-issues.png)
+
+### In the Protocol
```
Comment (mathiasbynens) [inline src/content/en/updates/2020/08/chrome-devtools-issues-tab.md:88]:
```suggestion

### In the protocol
```

File: src/content/en/updates/2020/08/chrome-devtools-issues-tab.md [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+For the browser side, we first looked into how console messages were handled, because their behavior is very similar to what we had in mind for issues. [CodeSearch](https://source.chromium.org/chromium/chromium/src/+/master:content/public/browser/console_message.h) is usually a good starting point for explorations like these. It allows you to search and explore the whole source code of the Chromium project online. That way, we learned about the implementation of console messages and could build up a [parallel, but more structured way](https://chromium-review.googlesource.com/c/chromium/src/+/1991507) around the requirements we collected for the issues. 
+
+The work here is especially challenging because of all the security implications we always have to keep in mind. The Chromium project goes a long way to separate things into different processes and have them only communicate through controlled communication channels to prevent information leaks. Issues may contain sensitive information, so we have to take care to not send that information to a part of the browser that shouldn't know about it.
+
+### In DevTools frontend
```
Comment (mathiasbynens) [inline src/content/en/updates/2020/08/chrome-devtools-issues-tab.md:73]: empty newline after headings:

```suggestion
### In DevTools frontend

```

File: src/content/en/updates/2020/08/chrome-devtools-issues-tab.md [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+For the Issues tab, we decided to add a new [event](https://chromedevtools.github.io/devtools-protocol/tot/Audits/#event-issueAdded) to the Audits domain that triggers whenever a new issue is encountered. To make sure that we can also report on issues that arise while DevTools is not yet opened, the Audits domain stores the most recent issues and dispatches them as soon as DevTools connects. DevTools then collects all those issues and aggregates them.
+
+The CDP also enables other protocol clients, such as [Puppeteer](https://pptr.dev/), to receive and process issues. We hope the [structured issue information](https://chromedevtools.github.io/devtools-protocol/tot/Audits/#type-InspectorIssue) sent over the CDP will enable and simplify integration into existing continuous integration infrastructure. This way, issues can be detected and fixed even before the page is deployed!
+
+## Future
```
Comment (mathiasbynens) [inline src/content/en/updates/2020/08/chrome-devtools-issues-tab.md:95]:
```suggestion
## Future

```

**Review by mathiasbynens — APPROVED** [review]
LGTM with some minor suggestions for consistency

**Comment by mathiasbynens** [comment]
A note on file names and URLs (non-actionable for now): while `src/content/en/updates/2020/08/chrome-devtools-issues-tab.md` makes sense for WebFu, once we migrate to a separate website that already has “Chrome DevTools” somewhere in its URL, we’ll want to rename the slug from `chrome-devtools-issues-tab` just `issues-tab`.

**Comment by mathiasbynens** [comment]
Oh, and one more thing: try and make sure you run the images through Squoosh or ImageOptim.app. On `related-issues.png` for example, we can save 4.9% of file size.

**Comment by chybie** [comment]
> Oh, and one more thing: try and make sure you run the images through Squoosh or ImageOptim.app. On `related-issues.png` for example, we can losslessly save 4.9% of file size.

Yes, I actually ran all the images through tinypng.com because it can somehow compress png smaller than Squoosh sometimes... (the team is notified).

**Comment by mathiasbynens** [comment]
> > Oh, and one more thing: try and make sure you run the images through Squoosh or ImageOptim.app. On `related-issues.png` for example, we can losslessly save 4.9% of file size.
> 
> Yes, I actually ran all the images through tinypng.com because it can somehow compress png smaller than Squoosh sometimes... (the team is notified).

Yeah, the same goes for `ImageOptim.app` with these settings:

**Comment by chybie** [comment]
> A note on file names and URLs (non-actionable for now): while `src/content/en/updates/2020/08/chrome-devtools-issues-tab.md` makes sense for WebFu, once we migrate to a separate website that already has “Chrome DevTools” somewhere in its URL, we’ll want to rename the slug from `chrome-devtools-issues-tab` just `issues-tab`.

Good one! I updated the link to `/web/updates/2020/08/devtools/issues-tab` for now instead. When the url is updated, I will strip off the folder instead. Preview: https://pr-8968-dot-web-central.appspot.com/web/updates/2020/08/devtools/issues-tab

**Comment by mathiasbynens** [comment]
> > A note on file names and URLs (non-actionable for now): while `src/content/en/updates/2020/08/chrome-devtools-issues-tab.md` makes sense for WebFu, once we migrate to a separate website that already has “Chrome DevTools” somewhere in its URL, we’ll want to rename the slug from `chrome-devtools-issues-tab` just `issues-tab`.
> 
> Good one! I updated the link to `/web/updates/2020/08/devtools/issues-tab` for now instead. When the url is updated, I will strip off the folder instead. Preview: https://pr-8968-dot-web-central.appspot.com/web/updates/2020/08/devtools/issues-tab

Hmm, I'm not sure that's better URL design, given that `https://pr-8968-dot-web-central.appspot.com/web/updates/2020/08/devtools` 404s. But then again, it seems WebFu URL design is broken anyway (e.g. `https://pr-8968-dot-web-central.appspot.com/web/updates/2020/08` doesn't work either).

**Comment by mathiasbynens** [comment]
Heh, it also seems to break the "What's New in DevTools (Chrome 86)" link in the sidebar.

**Pull request merged**